### PR TITLE
AE: updates

### PR DESCRIPTION
**Pull request by FernetMenta**
se title

@AchimTuran @AlwinEsch 
In order to make this compile, I commented some lines of ASDP, see "ADSP: temp changes for merge"

@garbear could you please adjust RetroPlayer to the changes

### CODE REVIEW

File: xbmc/ServiceManager.h
Comment (garbear) [inline xbmc/ServiceManager.h]: duplicate line

**Comment by AchimTuran** [comment]
@alwinesch
Can you have a look at the commented lines? I currently have no access to my PC. I can't see the issues on my phone from Github.

File: xbmc/ServiceManager.h
Comment (FernetMenta) [inline xbmc/ServiceManager.h]: thanks, I will fix

**Comment by wsnipex** [comment]
jenkins build this please

**Comment by FernetMenta** [comment]
@peak3d could you please check the Android related build failures?

**Comment by wsnipex** [comment]
jenkins build this please

**Comment by FernetMenta** [comment]
any takers for adapting the useless test drivers?

**Comment by wsnipex** [comment]
I tried to look at the tests, but this is over my paygrade.
It segfaults  here: https://github.com/FernetMenta/xbmc/blob/aeupdate/xbmc/ServiceManager.cpp#L181 so AE is not initialized.

File: xbmc/settings/dialogs/GUIDialogAudioDSPSettings.cpp
Comment (Paxxi) [inline xbmc/settings/dialogs/GUIDialogAudioDSPSettings.cpp]: Should this be removed? Why is it commented out?

File: xbmc/settings/dialogs/GUIDialogAudioDSPSettings.cpp
Comment (Paxxi) [inline xbmc/settings/dialogs/GUIDialogAudioDSPSettings.cpp]: Should this be removed? Why is it commented out?

File: xbmc/settings/dialogs/GUIDialogAudioDSPSettings.cpp
Comment (Paxxi) [inline xbmc/settings/dialogs/GUIDialogAudioDSPSettings.cpp]: Should this be removed? Why is it commented out?

File: xbmc/Application.cpp
Comment (Paxxi) [inline xbmc/Application.cpp]: don't use C-Style casts, see coding guidelines

File: xbmc/Application.cpp
Comment (Paxxi) [inline xbmc/Application.cpp]: don't use C-Style casts, see coding guidelines

File: xbmc/ServiceManager.cpp
Comment (Paxxi) [inline xbmc/ServiceManager.cpp]: is this needed? shouldn't it call Shutdown automatically in the destructor?

File: xbmc/addons/interfaces/AudioDSP/AddonCallbacksAudioDSP.cpp
Comment (Paxxi) [inline xbmc/addons/interfaces/AudioDSP/AddonCallbacksAudioDSP.cpp]: Is this an owning pointer?

File: xbmc/addons/interfaces/AudioDSP/AddonCallbacksAudioDSP.cpp
Comment (Paxxi) [inline xbmc/addons/interfaces/AudioDSP/AddonCallbacksAudioDSP.cpp]: don't use C-Style casts, see coding guidelines

File: xbmc/addons/interfaces/AudioEngine/AddonCallbacksAudioEngine.cpp
Comment (Paxxi) [inline xbmc/addons/interfaces/AudioEngine/AddonCallbacksAudioEngine.cpp]: Don't use C-Style casts, see coding guidelines

File: xbmc/cores/AudioEngine/Engines/ActiveAE/ActiveAE.cpp
Comment (Paxxi) [inline xbmc/cores/AudioEngine/Engines/ActiveAE/ActiveAE.cpp]: use unique_ptr

File: xbmc/cores/AudioEngine/Engines/ActiveAE/ActiveAESettings.cpp
Comment (Paxxi) [inline xbmc/cores/AudioEngine/Engines/ActiveAE/ActiveAESettings.cpp]: use a for each loop and move the firstDevice assignment out of the loop

File: xbmc/cores/AudioEngine/Engines/ActiveAE/ActiveAESound.cpp
Comment (Paxxi) [inline xbmc/cores/AudioEngine/Engines/ActiveAE/ActiveAESound.cpp]: use a reference to CActiveAE to make it explicit to the compiler and reader that it's valid for the lifetime of this class.

File: xbmc/cores/AudioEngine/Engines/ActiveAE/ActiveAESound.h
Comment (Paxxi) [inline xbmc/cores/AudioEngine/Engines/ActiveAE/ActiveAESound.h]: use a reference to CActiveAE to make it explicit to the compiler and reader that it's valid for the lifetime of this class.

File: xbmc/cores/AudioEngine/Engines/ActiveAE/ActiveAEStream.h
Comment (Paxxi) [inline xbmc/cores/AudioEngine/Engines/ActiveAE/ActiveAEStream.h]: use a reference to CActiveAE to make it explicit to the compiler and reader that it's valid for the lifetime of this class.

File: xbmc/cores/AudioEngine/Engines/ActiveAE/AudioDSPAddons/ActiveAEDSP.cpp
Comment (Paxxi) [inline xbmc/cores/AudioEngine/Engines/ActiveAE/AudioDSPAddons/ActiveAEDSP.cpp]: why is this commented out? should it be removed?

File: xbmc/cores/AudioEngine/Engines/ActiveAE/AudioDSPAddons/ActiveAEDSP.cpp
Comment (Paxxi) [inline xbmc/cores/AudioEngine/Engines/ActiveAE/AudioDSPAddons/ActiveAEDSP.cpp]: why is this commented out? should it be removed?

File: xbmc/cores/AudioEngine/Engines/ActiveAE/AudioDSPAddons/ActiveAEDSP.cpp
Comment (Paxxi) [inline xbmc/cores/AudioEngine/Engines/ActiveAE/AudioDSPAddons/ActiveAEDSP.cpp]: why is this commented out? should it be removed?

File: xbmc/cores/AudioEngine/Engines/ActiveAE/AudioDSPAddons/ActiveAEDSP.cpp
Comment (Paxxi) [inline xbmc/cores/AudioEngine/Engines/ActiveAE/AudioDSPAddons/ActiveAEDSP.cpp]: should be handled in the destructor of AE_DSP_ADDON so m_addonToDestroy.clear should be enough

File: xbmc/cores/AudioEngine/Engines/ActiveAE/AudioDSPAddons/ActiveAEDSP.cpp
Comment (Paxxi) [inline xbmc/cores/AudioEngine/Engines/ActiveAE/AudioDSPAddons/ActiveAEDSP.cpp]: check for nullptr

File: xbmc/cores/AudioEngine/Engines/ActiveAE/AudioDSPAddons/ActiveAEDSP.h
Comment (Paxxi) [inline xbmc/cores/AudioEngine/Engines/ActiveAE/AudioDSPAddons/ActiveAEDSP.h]: use a vector unless the intent is to do an enormous amount of insertions/removals, vectors are generally better for speed.

**Review by Paxxi — CHANGES_REQUESTED** [review]

File: xbmc/cores/AudioEngine/Engines/ActiveAE/AudioDSPAddons/ActiveAEDSP.h
Comment (AchimTuran) [inline xbmc/cores/AudioEngine/Engines/ActiveAE/AudioDSPAddons/ActiveAEDSP.h]: This is for speed, as it can happen during audio processing.

Comment (Paxxi) [inline xbmc/cores/AudioEngine/Engines/ActiveAE/AudioDSPAddons/ActiveAEDSP.h]: If it's for speed a vector may be a lot better, better cache locality and prefetching when reading from it. Either way works, just pointing it out

File: xbmc/cores/AudioEngine/Engines/ActiveAE/ActiveAESound.h
Comment (Paxxi) [inline xbmc/cores/AudioEngine/Engines/ActiveAE/ActiveAESound.h]: Actually, I spend some more time thinking about this and the design is completely broken. ActiveAESound should not know about ActiveAE or the other way around. Circular dependencies always suck and cause issues so this needs a redesign to be acceptable.

I would suggest making ActiveAESound and ActiveAEStream observable and have ActiveAE listen to their events. It's possibly an alternative to use the AE messaging system as well to to fix it.

Either way, this needs fixing before merge.

Comment (FernetMenta) [inline xbmc/cores/AudioEngine/Engines/ActiveAE/ActiveAESound.h]: I don't consider it broken

File: xbmc/cores/AudioEngine/Engines/ActiveAE/AudioDSPAddons/ActiveAEDSP.h
Comment (garbear) [inline xbmc/cores/AudioEngine/Engines/ActiveAE/AudioDSPAddons/ActiveAEDSP.h]: oh and the commit msg looks like it was c/ped without changing

File: xbmc/cores/AudioEngine/Engines/ActiveAE/ActiveAESound.h
Comment (FernetMenta) [inline xbmc/cores/AudioEngine/Engines/ActiveAE/ActiveAESound.h]: did I miss that @Paxxi claimed the maintainer role for ae?

Comment (Paxxi) [inline xbmc/cores/AudioEngine/Engines/ActiveAE/ActiveAESound.h]: There's basically no case where a circular dependency isn't broken. If you actually care about our architecture as you claim then you wouldn't accept this as a design going forward. 

This adds to the mess we're trying to clean up and get away from. 

I don't claim to be a maintainer of AE but I care about our overall code quality and try to fix/educate wherever possible.

Comment (FernetMenta) [inline xbmc/cores/AudioEngine/Engines/ActiveAE/ActiveAESound.h]: noted but this is not you decision here. AE is a rather self contained component and this code has no influence on anly other code outside AE. for this round I'll keep it as is

File: xbmc/cores/AudioEngine/Engines/ActiveAE/AudioDSPAddons/ActiveAEDSP.h
Comment (AchimTuran) [inline xbmc/cores/AudioEngine/Engines/ActiveAE/AudioDSPAddons/ActiveAEDSP.h]: Ah I was in a hurry. It is for insertions and removes.

@garbear
Thx. For the comment.

File: xbmc/cores/AudioEngine/Engines/ActiveAE/ActiveAESound.h
Comment (AchimTuran) [inline xbmc/cores/AudioEngine/Engines/ActiveAE/ActiveAESound.h]: @paxxi
> This adds to the mess we're trying to clean up and get away from.

WE dropped CAFactory and this dependency was already there. If we redesign it this PRS will get even bigger and worser to review. So please do not block this PR!

Comment (Paxxi) [inline xbmc/cores/AudioEngine/Engines/ActiveAE/ActiveAESound.h]: That is a perfectly valid reason. Can't do everything in one go but please consider revisiting this and fixing this sometime

Comment (FernetMenta) [inline xbmc/cores/AudioEngine/Engines/ActiveAE/ActiveAESound.h]: For the record. There is hardly any other component in Kodi that is self contained like AE. Even  AE related settings are owned by AE and are not accessed from the outside. In this iteration with killing the old factory it gets another improvement.

Calling AE broken in redard to the rest of the app feels somehow rediculous. In regards to dependencies the code I have maintained has improved a lot. Coming to my place and trying to reject code I wrote for code I wrote feels offending.

File: xbmc/ServiceManager.h
Comment (AchimTuran) [inline xbmc/ServiceManager.h]: thx @garbear 
should be fixed with this [PR](https://github.com/FernetMenta/xbmc/pull/420/files#diff-e0852c06443e72e1a8e07ce903e1406a).

File: xbmc/ServiceManager.cpp
Comment (AchimTuran) [inline xbmc/ServiceManager.cpp]: Yes it is needed. This is how AE's API is designed.

File: xbmc/addons/interfaces/AudioDSP/AddonCallbacksAudioDSP.cpp
Comment (AchimTuran) [inline xbmc/addons/interfaces/AudioDSP/AddonCallbacksAudioDSP.cpp]: @Paxxi 
Owning pointer from what?

File: xbmc/cores/AudioEngine/Engines/ActiveAE/ActiveAE.cpp
Comment (AchimTuran) [inline xbmc/cores/AudioEngine/Engines/ActiveAE/ActiveAE.cpp]: @Paxxi 
Why a unique_ptr?

File: xbmc/cores/AudioEngine/Engines/ActiveAE/ActiveAESettings.cpp
Comment (AchimTuran) [inline xbmc/cores/AudioEngine/Engines/ActiveAE/ActiveAESettings.cpp]: @FernetMenta 
Can you have a look on this?

File: xbmc/cores/AudioEngine/Engines/ActiveAE/ActiveAESound.h
Comment (AchimTuran) [inline xbmc/cores/AudioEngine/Engines/ActiveAE/ActiveAESound.h]: I guess @Paxxi has used to wrong words right?
Right @FernetMenta AE is one of the best codebases around in Kodi and so far I know @Paxxi only tries to improve our overall architecture.

File: xbmc/cores/AudioEngine/Engines/ActiveAE/AudioDSPAddons/ActiveAEDSP.h
Comment (AchimTuran) [inline xbmc/cores/AudioEngine/Engines/ActiveAE/AudioDSPAddons/ActiveAEDSP.h]: @Paxxi 
all your 
> why is this commented out? should it be removed?
statements are inside AudioDSP. As I'm currently rewriting all lines are needed as hints for me. I will add todos there and remove them during rewritting. Is this a valid way for you?

File: xbmc/cores/AudioEngine/Engines/ActiveAE/AudioDSPAddons/ActiveAEDSP.cpp
Comment (AchimTuran) [inline xbmc/cores/AudioEngine/Engines/ActiveAE/AudioDSPAddons/ActiveAEDSP.cpp]: will be removed

File: xbmc/cores/AudioEngine/Engines/ActiveAE/AudioDSPAddons/ActiveAEDSP.cpp
Comment (AchimTuran) [inline xbmc/cores/AudioEngine/Engines/ActiveAE/AudioDSPAddons/ActiveAEDSP.cpp]: No this is not enough. AudioDSP is a long lifetime object and it's add-ons can be enabled or disabled all the time. This will be required for binary add-on deployment. :-)

File: xbmc/cores/AudioEngine/Engines/ActiveAE/AudioDSPAddons/ActiveAEDSP.cpp
Comment (AchimTuran) [inline xbmc/cores/AudioEngine/Engines/ActiveAE/AudioDSPAddons/ActiveAEDSP.cpp]: @Paxxi 
Thx for the hint. I will check this for AudioDSP V2.0. As AudioDSP will be completely disabled I will avoid this check.

File: xbmc/cores/AudioEngine/Engines/ActiveAE/ActiveAESettings.cpp
Comment (FernetMenta) [inline xbmc/cores/AudioEngine/Engines/ActiveAE/ActiveAESettings.cpp]: just code cosmetics, ignore in this round

File: xbmc/cores/AudioEngine/Engines/ActiveAE/ActiveAE.cpp
Comment (Paxxi) [inline xbmc/cores/AudioEngine/Engines/ActiveAE/ActiveAE.cpp]: To avoid any missing delete

File: xbmc/addons/interfaces/AudioDSP/AddonCallbacksAudioDSP.cpp
Comment (Paxxi) [inline xbmc/addons/interfaces/AudioDSP/AddonCallbacksAudioDSP.cpp]: Is the method required to delete it or is that managed elsewhere is what I meant by owning pointer. Raw pointers are fine if they're not owning the resource, should be avoided if they're owning and instead use unique/shared_ptr

File: xbmc/cores/AudioEngine/Engines/ActiveAE/ActiveAESound.h
Comment (Paxxi) [inline xbmc/cores/AudioEngine/Engines/ActiveAE/ActiveAESound.h]: I did not speak of AE as a whole, I commented on this specific implementation and as I mentioned before, a circular dependency is basically never a valid design, this does not in any way depend on how good the overall code is. 

I also mentioned that I consider it acceptable if it's part of a redesign where the end goal gets rid of it as that can be necessary. 

I don't see why it matters who wrote this and the surrounding code. Are you trying to say that only the person who wrote the code is allowed to comment? That to me does not seem like a good way to work as a team and I'd rather avoid it.

Comment (FernetMenta) [inline xbmc/cores/AudioEngine/Engines/ActiveAE/ActiveAESound.h]: (almost) everybody can comment but the maintainer decides what is acceped and what is recected.

**Comment by FernetMenta** [comment]
jenkins build this please

**Comment by AchimTuran** [comment]
From my side this PR is finished. 

I will wait until Sunday and then merge it.

**Comment by popcornmix** [comment]
Could this be rebased?

**Comment by AchimTuran** [comment]
@FernetMenta 
Please rebase this PR.

**Comment by FernetMenta** [comment]
done
jenkind build this please

**Comment by FernetMenta** [comment]
jenkins build this please

**Comment by FernetMenta** [comment]
jenkins build this please

File: xbmc/pvr/PVRContextMenus.cpp
Comment (Rechi) [inline xbmc/pvr/PVRContextMenus.cpp]: `ServiceBroker.h` need to be included by this file
it was previously included through `ActiveAEDSP.h`

File: xbmc/cores/AudioEngine/Engines/ActiveAE/AudioDSPAddons/ActiveAEDSP.cpp
Comment (Rechi) [inline xbmc/cores/AudioEngine/Engines/ActiveAE/AudioDSPAddons/ActiveAEDSP.cpp]: `for (unsigned int i = 0;.. )`, because `m_usedProcessesCnt` is unsiged

**Comment by FernetMenta** [comment]
@Rechi done, thanks
jenkins build this please

**Comment by FernetMenta** [comment]
@AchimTuran undifined refs in adsp caused by changes in WindowMangaer, could you have a look?

**Comment by AchimTuran** [comment]
Sure.

File: xbmc/settings/dialogs/CMakeLists.txt
Comment (Rechi) [inline xbmc/settings/dialogs/CMakeLists.txt]: Removing `GUIDialogAudioDSPManager.cpp` from the build files causes the linking error.
After readding it (and commenting out some `CServiceBroker::GetADSP()`) it links again without errors.

Comment (FernetMenta) [inline xbmc/settings/dialogs/CMakeLists.txt]: thanks @Rechi 
I can fix this on Thursday if @AchimTuran isn't faster

Comment (AchimTuran) [inline xbmc/settings/dialogs/CMakeLists.txt]: @Rechi 
Thanks for the hints. The fixes are [here](https://github.com/FernetMenta/xbmc/pull/422).

**Comment by FernetMenta** [comment]
jenkins build this please

**Comment by FernetMenta** [comment]
jenkins build this please

**Comment by AchimTuran** [comment]
@FernetMenta The header files for the ServiceBroker and AE are missing in XBMCApp.cpp.

**Comment by FernetMenta** [comment]
probably but I don't have an Android dev environment to fix this. would be graet if an Android dev could provide a fix.

**Comment by peak3d** [comment]
send a PR to the dev branch for android issue

**Comment by FernetMenta** [comment]
thanks @peak3d 
jenkins build this please

**Comment by FernetMenta** [comment]
jenkins build this please

**Pull request merged**
